web/impact/impact/utils.py

```diff
@@ -66,3 +66,75 @@ def previous_instance(instance, query):
 
 def _find_instance(instance, query, order):
     return type(instance).objects.filter(query).order_by(order).first()
+
+"""
+return a use's program role grants grouped by the program family
+filter role grants by the provided user role or return
+all program role grant the user has ever had
+
+Time, Space Complexity
+O(n): time and space where n is the number of item in the result list
+"""
+def get_user_prg_role_by_program_family(user, user_roles=[]):
+    query = user.programrolegrant_set.filter(
+        program_role__user_role__isnull=False
+    )
+    if user_roles:
+        query = query.filter(
+            program_role__user_role__name__in=user_roles)
+    result = query.values_list(
+        'program_role__name', 'program_role__program__program_family__name')
+    prg_group = {}
+    for program_role_grant, program_family in result:
+        if prg_group.get(program_family):
+            prg_group[program_family].append(program_role_grant)
+        else:
+            prg_group[program_family] = [program_role_grant]
+    return prg_group
+
+"""
+fetch a the program status for all startup a given user
+belongs to
+
+Time, Space, Query Complexity
+Query: 0(n + 1) where n is the number of startup a user belongs to
+since for every startup in that list we are fetching the program status.
+and the +1 is for the query to fetch all startup for the user
+
+O(nm)Time/Space where n is the number of startup and m is
+the number of program_startup_status
+"""
+def get_user_startup_prg_role_by_program_family(user, startup_roles=[]):
+    startups = user.startup_set.all()
+    startup_status_group = {}
+    for startup in startups:
+        query = startup.program_startup_statuses()
+        if startup_roles:
+            query = query.filter(
+                startup_role__name__in=startup_roles
+            )
+        result = query.values_list("startup_status", "program__program_family__name")
+        for startup_status, program_family in result:
```

Comment:
consider refactoring this to one function. Similar block also found on `def get_user_prg_role_by_program_family`

Reply:
@iankigen I am not sure what you are referring to here, please say more

Reply:
[This block](https://github.com/masschallenge/impact-api/blob/b9c74bd88050804b17e17826d4532b279ecb39a8/web/impact/impact/utils.py#L88-L92) has similar functionality to [this](https://github.com/masschallenge/impact-api/blob/b9c74bd88050804b17e17826d4532b279ecb39a8/web/impact/impact/utils.py#L117-L121). Consider refactoring to one helper function

Reply:
I don't really see the additional benefit in abstracting those five lines, I think leaving it the way it is helps readability.

However I have done further abstraction on the query level

Reply:
@shanbady  @iankigen This is ready for final review